src/packbuilder.rs

```diff
@@ -0,0 +1,338 @@
+        }
+        Ok(())
+    }
+
+    /// Insert a commit object. This will add a commit as well as the completed
+    /// referenced tree.
+    pub fn insert_commit(&mut self, id: Oid) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_commit(self.raw, &*id.raw()));
+        }
+        Ok(())
+    }
+
+    /// Insert objects as given by the walk. Those commits and all objects they
+    /// reference will be inserted into the packbuilder.
+    pub fn insert_walk(&mut self, walk: &mut Revwalk) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_walk(self.raw, walk.raw()));
+        }
+        Ok(())
+    }
+
+    /// Recursively insert an object and its referenced objects. Insert the
+    /// object as well as any object it references.
+    pub fn insert_recursive(&mut self,
+                            id: Oid,
+                            name: Option<&str>)
+                            -> Result<(), Error> {
+        let name = try!(::opt_cstr(name));
+        unsafe {
+            try_call!(raw::git_packbuilder_insert_recur(self.raw,
+                                                        &*id.raw(),
+                                                        name));
+        }
+        Ok(())
+    }
+
+    /// Write the contents of the packfile to an in-memory buffer. The contents
+    /// of the buffer will become a valid packfile, even though there will be
+    /// no attached index.
+    pub fn write_buf(&mut self, buf: &mut Buf) -> Result<(), Error> {
+        unsafe {
+            try_call!(raw::git_packbuilder_write_buf(buf.raw(), self.raw));
+        }
+        Ok(())
+    }
+
+    /// Create the new pack and pass each object to the callback.
+    pub fn foreach(&mut self,
+                   mut cb: &mut ForEachCb)
+                   -> Result<(), Error> {
+        let ptr = &mut cb as *mut _;
+        unsafe {
+            try_call!(raw::git_packbuilder_foreach(self.raw,
+                                                   foreach_c,
+                                                   ptr as *mut _));
+        }
+        Ok(())
+    }
+
+    /// `progress` will be called with progress information during pack
+    /// building. Be aware that this is called inline with pack building
+    /// operations, so performance may be affected.
+    pub fn set_progress_callback<F>(&mut self, progress: F) -> Result<(), Error>
+        where F: FnMut(PackBuilderStage, u32, u32) -> bool + 'repo {
+        let ptr = Box::into_raw(Box::new(Box::new(progress) as Box<ProgressCb>));
+        let progress_c = Some(progress_c as raw::git_packbuilder_progress);
+        unsafe {
+            try_call!(raw::git_packbuilder_set_callbacks(self.raw,
+                                                         progress_c,
+                                                         ptr as *mut _));
+            self.progress = Some(Box::from_raw(ptr));
```

Comment:
I think it is, we need to pass in the raw pointer of the outer box as the callback argument and simultaneously keep a reference to it for cleanup when the `Packbuilder` is dropped. That reference in `self.progress` could be changed to a raw pointer as well and the `Box::from_raw` be called in `Packbuilder::drop` instead, this way just seemed cleaner to me.

Although, I just realised this may be breaking some aliasing rules, during a call to `extern fn progress_c` the inner box containing the callback is both simultaneously owned by the unborrowed box in the Packbuilder and referenced by a `*mut` in `progress_c`. Might be safer to store the raw pointer just to make sure there's no way this could be unsafe.